nack/receiver_interceptor.go

```diff
@@ -0,0 +1,174 @@
+package nack
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pion/interceptor"
+	"github.com/pion/logging"
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+)
+
+// ReceiverInterceptor interceptor generates nack messages.
+type ReceiverInterceptor struct {
+	interceptor.NoOp
+	size        uint16
+	skipLastN   uint16
+	interval    time.Duration
+	receiveLogs *sync.Map
+	m           sync.Mutex
+	wg          sync.WaitGroup
+	close       chan struct{}
+	log         logging.LeveledLogger
+}
+
+// NewReceiverInterceptor returns a new ReceiverInterceptor interceptor
+func NewReceiverInterceptor(size uint16, skipLastN uint16, interval time.Duration, log logging.LeveledLogger) (*ReceiverInterceptor, error) {
+	_, err := NewReceiveLog(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ReceiverInterceptor{
+		NoOp:        interceptor.NoOp{},
+		size:        size,
+		skipLastN:   skipLastN,
+		interval:    interval,
+		receiveLogs: &sync.Map{},
+		close:       make(chan struct{}),
+		log:         log,
+	}, nil
+}
+
+// BindRTCPWriter lets you modify any outgoing RTCP packets. It is called once per PeerConnection. The returned method
+// will be called once per packet batch.
+func (n *ReceiverInterceptor) BindRTCPWriter(writer interceptor.RTCPWriter) interceptor.RTCPWriter {
+	n.m.Lock()
+	select {
+	case <-n.close:
+		// already closed
+		n.m.Unlock()
+		return writer
+	default:
+	}
+
+	n.wg.Add(1)
+	n.m.Unlock()
+
+	go n.loop(writer)
+
+	return writer
+}
+
+// BindRemoteStream lets you modify any incoming RTP packets. It is called once for per RemoteStream. The returned method
+// will be called once per rtp packet.
+func (n *ReceiverInterceptor) BindRemoteStream(info *interceptor.StreamInfo, reader interceptor.RTPReader) interceptor.RTPReader {
+	hasNack := false
+	for _, fb := range info.RTCPFeedback {
+		if fb.Type == "nack" && fb.Parameter == "" {
+			hasNack = true
+		}
+	}
```

Comment:
There should be a utility function for that.

Reply:
Fixed! `streamSupportNack` in `pkg/interceptor`